File: backend/dslearn/src/main/java/com/mc656/dslearn/dtos/UserMetricsDTO.java
package com.mc656.dslearn.dtos;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.List;
import java.util.Map;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Builder
public class UserMetricsDTO {
    
    // Métricas gerais de tópicos
    private Integer totalTopics;
    private Integer completedTopics;
    private Integer inProgressTopics;
    private Integer notStartedTopics;
    private Double topicsCompletionPercentage;
    
    // Métricas gerais de exercícios
    private Integer totalExercises;
    private Integer completedExercises;
    private Integer inProgressExercises;
    private Integer notStartedExercises;
    private Double exercisesCompletionPercentage;
    
    // Métricas por dificuldade
    private Map<String, ExercisesByDifficultyMetrics> exercisesByDifficulty;
    
    // Métricas por tópico (quantidade de exercícios finalizados por tópico)
    private List<ExercisesByTopicMetrics> exercisesByTopic;
    
    // Progresso geral do usuário (0-100)
    private Double overallProgress;
    
    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    @Builder
    public static class ExercisesByDifficultyMetrics {
        private Integer total;
        private Integer completed;
        private Integer inProgress;
        private Integer notStarted;
        private Double completionPercentage;
    }
    
    @Data
    @AllArgsConstructor
    @NoArgsConstructor
    @Builder
    public static class ExercisesByTopicMetrics {
        private String topicName;
        private Integer totalExercises;
        private Integer completedExercises;
        private Integer inProgressExercises;
        private Integer notStartedExercises;
        private Double completionPercentage;
    }
}

File: app/DSLearn/app/src/main/java/com/unicamp/dslearn/data/datasource/remote/dto/UserMetricsDTO.kt
package com.unicamp.dslearn.data.datasource.remote.dto


import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class UserMetricsDTO(
    @SerialName("totalTopics") val totalTopics: Int,
    @SerialName("completedTopics") val completedTopics: Int,
    @SerialName("inProgressTopics") val inProgressTopics: Int,
    @SerialName("notStartedTopics") val notStartedTopics: Int,
    @SerialName("topicsCompletionPercentage") val topicsCompletionPercentage: Float,
    @SerialName("totalExercises") val totalExercises: Int,
    @SerialName("completedExercises") val completedExercises: Int,
    @SerialName("inProgressExercises") val inProgressExercises: Int,
    @SerialName("notStartedExercises") val notStartedExercises: Int,
    @SerialName("exercisesCompletionPercentage") val exercisesCompletionPercentage: Float,
    @SerialName("exercisesByDifficulty")
    val exercisesByDifficulty: ExercisesByDifficultyDTO,
    @SerialName("exercisesByTopic")
    val exercisesByTopic: List<ExercisesByTopicDTO> = emptyList(),
    @SerialName("overallProgress") val overallProgress: Float
)

@Serializable
data class ExercisesByDifficultyDTO(
    @SerialName("Easy") val easy: DifficultyStatsDTO,
    @SerialName("Medium") val medium: DifficultyStatsDTO,
    @SerialName("Hard") val hard: DifficultyStatsDTO
)

@Serializable
data class DifficultyStatsDTO(
    @SerialName("total") val total: Int,
    @SerialName("completed") val completed: Int,
    @SerialName("inProgress") val inProgress: Int,
    @SerialName("notStarted") val notStarted: Int,
    @SerialName("completionPercentage") val completionPercentage: Float
)

@Serializable
data class ExercisesByTopicDTO(
    @SerialName("topicName") val topicName: String,
    @SerialName("totalExercises") val totalExercises: Int,
    @SerialName("completedExercises") val completedExercises: Int,
    @SerialName("inProgressExercises") val inProgressExercises: Int,
    @SerialName("notStartedExercises") val notStartedExercises: Int,
    @SerialName("completionPercentage") val completionPercentage: Float
)
